CHANGELOG.md

```diff
@@ -0,0 +1,167 @@
+  - `Bundle::try_map_authorization`
+  - `TryMapAuth`
+  - `impl {MapAuth, TryMapAuth} for (FnMut, FnMut, FnMut, FnMut)`
+    helpers to enable calling `Bundle::{map_authorization, try_map_authorization}`
+    with a set of closures.
+  - `testing` module, containing the following functions moved from
+    `zcash_primitives::transaction::components::sapling::testing`:
+    - `arb_output_description`
+    - `arb_bundle`
+- `sapling_crypto::circuit` module (moved from `zcash_proofs::circuit::sapling`).
+  Additional additions to this module:
+  - `{SpendParameters, OutputParameters}`
+  - `{SpendVerifyingKey, PreparedSpendVerifyingKey}`
+  - `{OutputVerifyingKey, PreparedOutputVerifyingKey}`
+- `sapling_crypto::constants` module.
+- `sapling_crypto::keys`:
+  - `SpendAuthorizingKey`
+  - `SpendValidatingKey`
+- `sapling_crypto::note_encryption`:
+  - `CompactOutputDescription` (moved from
+    `zcash_primitives::transaction::components::sapling`).
+  - `SaplingDomain::new`
+  - `Zip212Enforcement`
+- `sapling_crypto::prover::{SpendProver, OutputProver}`
+- `sapling_crypto::tree::Node::{from_bytes, to_bytes}`
+- `sapling_crypto::value`:
+  - `ValueCommitTrapdoor::from_bytes`
+  - `impl Sub<TrapdoorSum> for TrapdoorSum`
+  - `impl Sub<CommitmentSum> for CommitmentSum`
+- `sapling_crypto::zip32` module (moved from `zcash_primitives::zip32::sapling`).
+- `impl Debug for sapling_crypto::keys::{ExpandedSpendingKey, ProofGenerationKey}`
+- Test helpers, behind the `test-dependencies` feature flag:
+  - `sapling_crypto::prover::mock::{MockSpendProver, MockOutputProver}`
+
+### Changed
+- `sapling_crypto`:
+  - `BatchValidator::validate` now takes the `SpendVerifyingKey` and
+    `OutputVerifyingKey` newtypes.
+  - `SaplingVerificationContext::new` now always creates a context with ZIP 216
+    rules enforced, and no longer has a boolean for configuring this.
+  - `SaplingVerificationContext::{check_spend, final_check}` now use the
+    `redjubjub` crate types for `rk`, `spend_auth_sig`, and `binding_sig`.
+  - `SaplingVerificationContext::{check_spend, check_output}` now take
+    the `PreparedSpendVerifyingKey` and `PreparedOutputVerifyingKey`
+    newtypes.
+  - `SaplingVerificationContext::final_check` now takes its `value_balance`
+    argument as `V: Into<i64>` instead of `Amount`.
+- `sapling_crypto::address::PaymentAddress::create_note` now takes its `value`
+  argument as a `NoteValue` instead of as a bare `u64`.
+- `sapling_crypto::builder`:
+  - `SaplingBuilder` no longer has a `P: consensus::Parameters` type parameter.
+  - `SaplingBuilder::new` now takes a `Zip212Enforcement` argument instead of a
+    `P: consensus::Parameters` argument and a target height.
+  - `SaplingBuilder::add_spend` now takes `extsk` by reference. Also, it no
+    longer takes a `diversifier` argument as the diversifier may be obtained
+    from the note.
+  - `SaplingBuilder::add_output` now takes an `Option<[u8; 512]>` memo instead
```

Comment:
I wonder whether it would actually be more correct to take `[u8; 512]` rather than an `Option`al value; if we're taking the `Option` here, it means that we're enforcing the Zcash network `0xf6` convention on all users.

Reply:
The same comment applies to the Orchard builder.